server/src/main/java/com/cloud/network/Ipv6AddressManagerImpl.java

```diff
@@ -260,4 +173,37 @@ protected boolean isIp6Taken(Network network, String requestedIpv6) {
         return ip6Vo != null || nicSecondaryIpVO != null;
     }
 
+    @Override
+    public void setNicIp6Address(final NicProfile nic, final DataCenter dc, final Network network) {
```

Comment:
Sorry, if this is silly, but that is something I see a problem with.  You might be able to clarify this for me. I do not have a deep network knowledge, and that is probably why I am seeing this problem.

My biggest doubt lies here:
> Using Router Advertisements the routers advertise which subnet can be used and based on the MAC a client in that network will generate a stable address.

So, I am assuming that when a client receives the router advertisement, it will always generate the same IP address because this process is based on the MAC address (I am assuming a deterministic approach here). If that is not the case, then you can discard the scenario that follows.

Let’s say IP `A` is generated for network `n` by a `client X` that has `MAC Y`.

Now, let’s pretend `client x` has not been deployed yet, but we have deployed `client Z`. This client has `MAC T`, and based on the network `n` that has been advertised, it (`client Z`) generates IP `B`. Afterwards, some user/operator can assign IP `A` to `client Z`. What happens if we deploy `client x` now?

I ask that because, when you call this method, and if the process of `NetUtils.EUI64Address` is deterministic, you might generate an IP address that is already in use by some other VM. Would it be possible?

Reply:
@rafaelweingartner A RA (Router Advertisements) is send out every X seconds by all IPv6 routers in a network.

All Instances (clients!) receive this RA and generate this address.

See this output below, this a IPv6 VM I have running:

<pre>
root@drive:~# ip -6 addr show dev ens3
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 2a00:f10:121:400:428:daff:fe00:49e/64 scope global mngtmpaddr dynamic 
       valid_lft 2591988sec preferred_lft 604788sec
    inet6 fe80::428:daff:fe00:49e/64 scope link 
       valid_lft forever preferred_lft forever
root@drive:~# ip -6 r
2a00:f10:121:400::/64 dev ens3  proto kernel  metric 256  expires 2591999sec pref medium
fe80::/64 dev ens3  proto kernel  metric 256  pref medium
default via fe80::21d:b502:c12e:e029 dev ens3  proto ra  metric 1024  expires 58sec hoplimit 64 pref medium
default via fe80::ae4b:c802:c180:9029 dev ens3  proto ra  metric 1024  expires 59sec hoplimit 64 pref medium
root@drive:~# 
</pre>

See how the IPv6 address 'matches' the MAC Address.

Each Instance has a unique IP address. So every time you deploy an Instance CloudStack will generate a new MAC and thus calculate a different IPv6 address for that Instance.

So even though if you don't start the first Instance yet, a new Instance will never get the same IPv6 address since the address is based on the MAC which is already unique.

So yes, the address of an Instance is always stable, but a conflict is impossible as each Instance has it's own MAC and that also makes the IPv6 address unique.

This already works in Basic Networking. I'm just bringing the functionality to Advanced Networking.

Reply:
Ok, I am almost seeing the full picture, but I still have a doubt.

For instance, in the example you presented. The IPv6 `2a00:f10:121:400:428:daff:fe00:49e` was "assigned"/"generate" to this `drive` machine based on its MAC address. What would happen, if before the deployment of the `drive` machine, we manually assign the address `2a00:f10:121:400:428:daff:fe00:49e` to a different machine, that has a different MAC address.

Reply:
Just an update for you, I am referring to cases where we have multiple IPs to the same NIC. The NIC has only one MAC. Therefore, the other IPs will not be based on the NIC MAC address.

The question I am raising is regarding these other IPs, and if they might represent the IP of some valid MAC address that might be used in the same subnet. Therefore, when a NIC with this MAC address (that was not in use before, but might be used in the future) is deployed/created in the same subnet, what happens if the IP that is generate is already assigned to other NIC manually?

Reply:
@rafaelweingartner I see. What we should do is that if a user wants to manually specify an IPv6 address for the NIC (as a Secondary IP) should not be based on SLAAC.

The thing I can think of is during the EUI-64 I could double check if no NIC or secondary IPv6 address has the IPv6. In such a case the deployment should fail.

The other thing we can do is when a user wants to add a secondary IPv6 address we check if that address is a EUI-64 address and then refuse it.

But this would be outside the scope of this PR I think as this isn't about secondary IPv6 addresses.

Keep in mind, Basic Networking currently works this way and Advanced Networking doesn't have any IPv6 support.

Reply:
@rafaelweingartner: https://community.cisco.com/t5/networking-documents/understanding-ipv6-eui-64-bit-address/ta-p/3116953

<pre>
Extended Unique Identifier (EUI), as per RFC2373, allows a host to assign iteslf a unique 64-Bit IP Version 6 interface identifier (EUI-64). This feature is a key benefit over IPv4 as it eliminates the need of manual configuration or DHCP as in the world of IPv4. The IPv6 EUI-64 format address is obtained through the 48-bit MAC address. The MAC address is first separated into two 24-bits, with one being OUI (Organizationally Unique Identifier) and the other being NIC specific. The 16-bit 0xFFFE is then inserted between these two 24-bits for the 64-bit EUI address. IEEE has chosen FFFE as a reserved value which can only appear in EUI-64 generated from the an EUI-48 MAC address.
</pre>

We could check (in another PR) that if a user wants to specify it's own IPv6 address it should not be EUI-64 by checking if *FFFE* is present in the address.